fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/clientlog/KeySpaceCountTree.java

```diff
@@ -83,35 +389,181 @@ protected TupleKeyCountTree newPrefixChild(@Nonnull Object prefix) {
+                        } catch (RecordCoreException ex) {
+                            break;
+                        }
+                        return CompletableFuture.completedFuture(new ResolvedIndexKeyspace(resolvedParent, index));
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+        if (recordTypeKeyspace != null) {
+            KeyExpression primaryKey = recordTypeKeyspace.getRecordType().getPrimaryKey();
+            if (distance + 1 < primaryKey.getColumnSize()) {
+                return resolvePrimaryKeyField(context, resolvedParent, object, primaryKey, distance + 1);
+            }
         }
+        if (indexKeyspace != null &&
+                indexKeyspace.getParent() instanceof ResolvedRecordStoreKeyspace &&
+                ((ResolvedRecordStoreKeyspace)indexKeyspace.getParent()).getRecordStoreKeyspace() == FDBRecordStoreKeyspace.INDEX) {
+            Index index = indexKeyspace.getIndex();
+            KeyExpression indexExpression = index.getRootExpression();
+            if (distance < indexExpression.getColumnSize()) {
+                return resolveIndexField(context, resolvedParent, object, index, indexExpression, distance);
+            }
+        }
+        return UNRESOLVED;
+    }
+
+    /**
+     * Determine whether this leaf of the {@link KeySpaceDirectory} tree is the root of a {@link com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore}.
+     * @param context an open transaction to use to read from the database
+     * @param resolvedParent the resolved parent node
+     * @param object the {@link com.apple.foundationdb.tuple.Tuple} element for this node
+     * @return {@code true} if this path stores a record store
+     */
+    protected boolean isRecordStoreLeaf(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object) {
+        return false;
+    }
+
+    /**
+     * Given a key space path for which {@link #isRecordStoreLeaf} is {@code true}, get the record store's meta-data.
+     * @param context an open transaction to use to read from the database
+     * @param resolvedParent the resolved parent node
+     * @param object the {@link com.apple.foundationdb.tuple.Tuple} element for this node
+     * @return a future that completes to the record store's meta-data or {@code null}
+     */
+    protected CompletableFuture<RecordMetaData> getRecordStoreMetaData(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object) {
+        return NO_META_DATA;
+    }
+
+    protected CompletableFuture<Resolved> resolvePrimaryKeyField(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object,
+                                                                 @Nonnull KeyExpression primaryKey, int index) {
+        List<KeyExpression> keys = primaryKey.normalizeKeyForPositions();
```

Comment:
Hm, we have a few places that we try and rendezvous the tuple we see with a key expression, but they're all a little scattershot. I think having this is here is probably fine, though perhaps when some of the new planner changes make that easier, it would be good to switch to that, or something.

Reply:
Yes, agreed. But that feels outside the scope of this branch.